pkg/binlog/reader/util.go

```diff
@@ -0,0 +1,135 @@
+				if err != nil {
+					return nil, terror.Annotatef(err, "update GTID set %v with GTID %s", latestGSet, nextGTIDStr)
+				}
+				latestPos = e.Header.LogPos
+			}
+		case *replication.XIDEvent:
+			if latestGSet == nil {
+				// GTID not enabled, can't get GTIDs for the position.
+				return nil, errors.Errorf("should have a GTIDEvent before the XIDEvent %+v", e.Header)
+			}
+			err = latestGSet.Update(nextGTIDStr)
+			if err != nil {
+				return nil, terror.Annotatef(err, "update GTID set %v with GTID %s", latestGSet, nextGTIDStr)
+			}
+			latestPos = e.Header.LogPos
+		case *replication.GTIDEvent:
+			if latestGSet == nil {
+				return nil, errors.Errorf("should have a PreviousGTIDsEvent before the GTIDEvent %+v", e.Header)
+			}
+			// learn from: https://github.com/siddontang/go-mysql/blob/c6ab05a85eb86dc51a27ceed6d2f366a32874a24/replication/binlogsyncer.go#L736
+			u, _ := uuid.FromBytes(ev.SID)
+			nextGTIDStr = fmt.Sprintf("%s:%d", u.String(), ev.GNO)
+		case *replication.MariadbGTIDEvent:
+			if latestGSet == nil {
+				return nil, errors.Errorf("should have a MariadbGTIDListEvent before the MariadbGTIDEvent %+v", e.Header)
+			}
+			// learn from: https://github.com/siddontang/go-mysql/blob/c6ab05a85eb86dc51a27ceed6d2f366a32874a24/replication/binlogsyncer.go#L745
+			GTID := ev.GTID
+			nextGTIDStr = fmt.Sprintf("%d-%d-%d", GTID.DomainID, GTID.ServerID, GTID.SequenceNumber)
+		case *replication.PreviousGTIDsEvent:
+			// if GTID enabled, we can get a PreviousGTIDEvent after the FormatDescriptionEvent
+			// ref: https://github.com/mysql/mysql-server/blob/8cc757da3d87bf4a1f07dcfb2d3c96fed3806870/sql/binlog.cc#L4549
+			// ref: https://github.com/mysql/mysql-server/blob/8cc757da3d87bf4a1f07dcfb2d3c96fed3806870/sql/binlog.cc#L5161
+			var gSet gtid.Set
+			gSet, err = gtid.ParserGTID(gmysql.MySQLFlavor, ev.GTIDSets)
+			if err != nil {
+				return nil, err
+			}
+			latestGSet = gSet.Origin()
+			flavor = gmysql.MySQLFlavor
+		case *replication.MariadbGTIDListEvent:
+			// a MariadbGTIDListEvent logged in every binlog to record the current replication state if GTID enabled
+			// ref: https://mariadb.com/kb/en/library/gtid_list_event/
+			gSet, err2 := event.GTIDsFromMariaDBGTIDListEvent(e)
+			if err2 != nil {
+				return nil, terror.Annotatef(err2, "get GTID set from MariadbGTIDListEvent %+v", e.Header)
+			}
+			latestGSet = gSet.Origin()
+			flavor = gmysql.MariaDBFlavor
+		}
+
+		if latestPos == endPos.Pos {
+			// reach the end position, return the GTID sets.
+			if latestGSet == nil {
+				return nil, errors.Errorf("no GTIDs get for position %s", endPos)
+			}
+			var latestGTIDs gtid.Set
+			latestGTIDs, err = gtid.ParserGTID(flavor, latestGSet.String())
+			if err != nil {
+				return nil, terror.Annotatef(err, "parse GTID set %s with flavor %s", latestGSet.String(), flavor)
+			}
+			return latestGTIDs, nil
+		} else if latestPos > endPos.Pos {
+			return nil, errors.Errorf("no GTIDs get for position %s", endPos)
```

Comment:
not understand, seems error message should be "not a complete GTIDs at position"?

Reply:
updated https://github.com/pingcap/dm/pull/858/commits/00da495d8b4e522ef929c7e0dbf81df8cb74752e. this often caused by an invalid position.